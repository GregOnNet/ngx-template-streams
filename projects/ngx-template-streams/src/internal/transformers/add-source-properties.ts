import { tsquery } from '@phenomnomnominal/tsquery';
import * as ts from 'typescript';
import { getComponentClass } from '../../utils/ast-helpers';
import { INTERNAL_PREFIX } from '../constants';
import { OBSERVABLE_EVENT_QUERY } from './queries';

export function addSourcePropertiesTransformer(context: ts.TransformationContext) {
  return (sourceFile: ts.SourceFile) => {
    return getComponentClass(sourceFile)
      .map(addSourceProperties(sourceFile, context))
      .getOrElse(sourceFile);
  };
}

function addSourceProperties(sourceFile: ts.SourceFile, context: ts.TransformationContext) {
  return (nodes: Array<ts.Node>) => {
    const visitor: ts.Visitor = (node: ts.Node) => {
      if (nodes.includes(node)) {
        const classDeclaration = node as ts.ClassDeclaration;

        const observableEvents = tsquery(classDeclaration, OBSERVABLE_EVENT_QUERY) as ts.ClassElement[];

        if (!observableEvents.length) {
          return node;
        }

        const templateStreamProperties = observableEvents.map((observableEvent: ts.PropertyDeclaration) => {
          const propertyName = observableEvent.name as ts.Identifier;

          return ts.createProperty(
            undefined,
            undefined,
            ts.createIdentifier(INTERNAL_PREFIX + propertyName.text),
            undefined,
            undefined,
            undefined
          );
        });

        return ts.updateClassDeclaration(
          classDeclaration,
          classDeclaration.decorators,
          classDeclaration.modifiers,
          classDeclaration.name,
          classDeclaration.typeParameters,
          classDeclaration.heritageClauses,
          ts.createNodeArray([...classDeclaration.members, ...templateStreamProperties])
        );
      }

      return ts.visitEachChild(node, visitor, context);
    };

    return ts.visitNode(sourceFile, visitor);
  };
}
