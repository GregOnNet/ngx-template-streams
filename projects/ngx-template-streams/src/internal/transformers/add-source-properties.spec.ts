import { resolve } from 'path';
import { fixtures, getFixtureFactory } from '../../testing/test-helpers';
import { transform } from '../../testing/ts-compiler';
import * as transformer from './add-source-properties';

const getComponentFixture = getFixtureFactory('component');
const getFixture = getFixtureFactory();

describe('AddSourcePropertiesTransformer', () => {
  let transformerSpy: jest.SpyInstance;

  beforeEach(() => {
    transformerSpy = jest.spyOn(transformer, 'addSourcePropertiesTransformer');
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should not apply transformation if the file is empty', () => {
    const { sourceFile } = getFixture('empty-file');

    const output = transform(resolve(fixtures, sourceFile), [transformer.addSourcePropertiesTransformer]);

    expect(output).toMatchSnapshot();
    expect(transformerSpy).toHaveBeenCalled();
  });

  it('should add source properties for each observable event', () => {
    const { sourceFile } = getComponentFixture('full-example');

    const output = transform(resolve(fixtures, sourceFile), [transformer.addSourcePropertiesTransformer]);

    expect(output).toMatchSnapshot();
  });

  it('should return original source file if it has no observable events', () => {
    const { sourceFile } = getComponentFixture('empty');

    const output = transform(resolve(fixtures, sourceFile), [transformer.addSourcePropertiesTransformer]);

    expect(output).toMatchSnapshot();
  });

  it('should transform multiple classes in a single file', () => {
    const { sourceFile } = getFixture('multiple-classes');

    const output = transform(resolve(fixtures, sourceFile), [transformer.addSourcePropertiesTransformer]);

    expect(output).toMatchSnapshot();
  });
});
