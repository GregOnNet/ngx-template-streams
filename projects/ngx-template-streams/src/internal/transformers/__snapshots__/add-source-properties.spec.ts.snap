// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addSourcePropertiesTransformer should add source properties for each observable event 1`] = `
"import { Component } from '@angular/core';
import { ObservableEvent } from 'ngx-template-streams';
import { Observable } from 'rxjs';
@Component({
    selector: 'test',
    template: \`
    <button (*click)=\\"click$\\">Click Me</button>
    <input (*focus)=\\"focus$\\" />
  \`,
    styleUrls: ['./app.component.css']
})
export class TestComponent {
    @ObservableEvent()
    click$: Observable<any>;
    @ObservableEvent()
    focus$: Observable<any>;
    __click$;
    __focus$;
}
"
`;

exports[`addSourcePropertiesTransformer should not apply transformation if the file is empty 1`] = `""`;

exports[`addSourcePropertiesTransformer should return original source file if it has no observable events 1`] = `
"import { Component } from '@angular/core';
@Component({
    selector: 'test',
    template: 'test',
    styleUrls: ['./empty.component.css']
})
export class TestComponent {
}
"
`;

exports[`addSourcePropertiesTransformer should transform multiple classes in a single file 1`] = `
"import { Component } from '@angular/core';
import { ObservableEvent } from 'ngx-template-streams';
import { Observable } from 'rxjs';
@Component({
    selector: 'a',
    template: '<button (*click)=\\"a$\\">Test<button>'
})
export class ClassA {
    @ObservableEvent()
    a$: Observable<any>;
    __a$;
}
@Component({
    selector: 'b',
    template: '<button (*click)=\\"b$\\">Test<button>'
})
export class ClassB {
    @ObservableEvent()
    b$: Observable<any>;
    __b$;
}
"
`;
